homeassistant/components/edl21/sensor.py

```diff
@@ -0,0 +1,158 @@
+"""Support for EDL21 Smart Meters."""
+
+from datetime import timedelta
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import STATE_UNKNOWN
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import Optional
+from homeassistant.util import Throttle
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = "edl21"
+CONF_SERIAL_PORT = "serial_port"
+ICON_POWER = "mdi:flash"
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=60)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({vol.Required(CONF_SERIAL_PORT): cv.string})
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the EDL21 sensor."""
+    hass.data[DOMAIN] = EDL21(hass, config, async_add_entities)
+    await hass.data[DOMAIN].connect()
+
+
+class EDL21:
+    """EDL21 handles telegrams sent by a compatible smart meter."""
+
+    def __init__(self, hass, config, async_add_entities) -> None:
+        """Initialize an EDL21 object."""
+        from sml.asyncio import SmlProtocol
+
+        self._cache = {}
+        self._hass = hass
+        self._async_add_entities = async_add_entities
+        self._proto = SmlProtocol(config[CONF_SERIAL_PORT])
+        self._proto.add_listener(self.event, ["SmlGetListResponse"])
+
+    async def connect(self):
+        """Connect to an EDL21 reader."""
+        await self._proto.connect(self._hass.loop)
+
+    def event(self, message_body) -> None:
+        """Handle events from pysml."""
+        from sml import SmlGetListResponse
+
+        assert isinstance(message_body, SmlGetListResponse)
+
+        new_devices = []
+        for telegram in message_body.get("valList", []):
+            obis = telegram.get("objName")
+            if not obis:
+                continue
+
+            device = self._cache.get(obis)
+            if not device:
+                device = self._cache[obis] = EDL21Entity(obis, telegram)
+                new_devices.append(device)
+            elif device.update_telegram(telegram):
+                self._hass.async_create_task(device.async_update_ha_state())
+
+        if new_devices:
+            self._hass.async_add_job(
+                self._async_add_entities(new_devices, update_before_add=True)
+            )
+
+
+class EDL21Entity(Entity):
+    """Entity reading values from EDL21 telegram."""
+
+    # OBIS format: A-B:C.D.E*F
+    _OBIS_NAMES = {
```

Comment:
Are these all the possible sensors? 

In that case, should we filter out Manufacturer and public key ?

Reply:
These are the ones used by the device I tested. In theory, there are many more sensors, but I expect most of them to never appear in the wild.

Why do you think we should filter out these two sensors? Or do you suggest filtering all manufacturer specific sensors (i.e. those starting with "129-")? The namespace is big enough to have no or only few collisions between the small number of manufacturers, I think.

Reply:
We should only provide sensors that make sense to the user. The manufacturer of a sensor is not a sensor value, it's something that is returned from device info.

Reply:
There's no guarantee for this information to be present on other devices. We could filter out this ID (129-129:199.130.3*255), but for other manufacturers, other IDs may be used for static information inside the vendor specific namespace. On the other hand, this namespace might as well include non-standard sensor values.

We could use _OBIS_NAMES as a whitelist, but it would make it harder for people to include new sensors in their user interface.

Reply:
What about using it as a whitelist, and then issue a one time warning whenever an unknown sensor is observed, to open an issue so that it can be added. That way we can fix each value once for each user. 

And then we would also have some values that we choose ot ignore, like Manufacturer.

Reply:
I implemented your proposal in my latest commit. Please review!